File: src/studica_control/src/components/examples/cpp/parsec_example.cpp
/*
 * parsec_example.cpp
 *
 * Subscribe to multi-zone ToF readings published by the Parsec component.
 * Prints nearest valid distance and a compact zone grid for 100 frames, then exits.
 * Polls get_min_distance via service every 5 seconds.
 *
 * Run:      ros2 run studica_control parsec_example
 * Requires: studica_launch.py running, parsec enabled in params.yaml
 *           sensors: ["parsec"]  (name must match SENSOR below)
 *
 * Topics subscribed:
 *   /parsec/zones     (studica_control/ParsecZoneMsg)
 *   /parsec/min_range (sensor_msgs/Range)
 * Service: /parsec/parsec_cmd (studica_control/SetData)
 *   Commands: get_config, get_zone_distance, get_min_distance
 */

#include <cmath>
#include <cstdio>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/range.hpp"
#include "studica_control/msg/parsec_zone_msg.hpp"
#include "studica_control/srv/set_data.hpp"

using namespace std::chrono_literals;

static const std::string SENSOR = "parsec";
static constexpr int kMaxPrints = 100;

class ParsecExample : public rclcpp::Node {
public:
    ParsecExample() : Node("parsec_example") {
        zones_sub_ = create_subscription<studica_control::msg::ParsecZoneMsg>(
            "/" + SENSOR + "/zones", 10,
            std::bind(&ParsecExample::on_zones, this, std::placeholders::_1));

        range_sub_ = create_subscription<sensor_msgs::msg::Range>(
            "/" + SENSOR + "/min_range", 10,
            std::bind(&ParsecExample::on_min_range, this, std::placeholders::_1));

        client_ = create_client<studica_control::srv::SetData>("/" + SENSOR + "/parsec_cmd");
        poll_timer_ = create_wall_timer(5s, std::bind(&ParsecExample::poll_service, this));

        RCLCPP_INFO(get_logger(), "listening on /%s/zones and /%s/min_range",
                    SENSOR.c_str(), SENSOR.c_str());
    }

private:
    static bool is_valid_mm(int16_t d) { return d >= 1 && d <= 4000; }

    void on_zones(const studica_control::msg::ParsecZoneMsg::SharedPtr msg) {
        const int side = (msg->zones == 64) ? 8 : 4;
        const int center = msg->zones / 2;
        const int16_t center_mm =
            (center < static_cast<int>(msg->fdist.size())) ? msg->fdist[static_cast<size_t>(center)] : -2;

        RCLCPP_INFO(get_logger(), "%3d | zones seq=%u count=%u center=%d mm",
                    print_count_, msg->seq, msg->zones, center_mm);

        for (int row = 0; row < side; ++row) {
            std::string line = "  ";
            for (int col = 0; col < side; ++col) {
                const size_t idx = static_cast<size_t>(row * side + col);
                if (idx >= msg->fdist.size()) {
                    break;
                }
                const int16_t d = msg->fdist[idx];
                char cell[16];
                if (d == -1) {
                    std::snprintf(cell, sizeof(cell), "%7s", "off");
                } else if (d == -2 || !is_valid_mm(d)) {
                    std::snprintf(cell, sizeof(cell), "%7s", "---");
                } else {
                    std::snprintf(cell, sizeof(cell), "%4d mm", d);
                }
                line += cell;
                if (col + 1 < side) {
                    line += ' ';
                }
            }
            RCLCPP_INFO(get_logger(), "%s", line.c_str());
        }

        ++print_count_;
        if (print_count_ >= kMaxPrints) {
            RCLCPP_INFO(get_logger(), "printed %d frames, exiting", kMaxPrints);
            rclcpp::shutdown();
        }
    }

    void on_min_range(const sensor_msgs::msg::Range::SharedPtr msg) {
        if (std::isinf(msg->range)) {
            RCLCPP_WARN(get_logger(), "no valid zones in frame");
        } else {
            RCLCPP_INFO(get_logger(), "nearest valid zone: %.3f m", msg->range);
        }
    }

    void poll_service() {
        if (!client_->wait_for_service(1s)) {
            RCLCPP_WARN(get_logger(), "parsec_cmd service not available");
            return;
        }
        auto req = std::make_shared<studica_control::srv::SetData::Request>();
        req->params = "get_min_distance";
        client_->async_send_request(req,
            [this](rclcpp::Client<studica_control::srv::SetData>::SharedFuture f) {
                const auto resp = f.get();
                if (resp->success) {
                    RCLCPP_INFO(get_logger(), "service get_min_distance: %s mm", resp->message.c_str());
                } else {
                    RCLCPP_WARN(get_logger(), "service get_min_distance failed: %s",
                                resp->message.c_str());
                }
            });
    }

    rclcpp::Subscription<studica_control::msg::ParsecZoneMsg>::SharedPtr zones_sub_;
    rclcpp::Subscription<sensor_msgs::msg::Range>::SharedPtr range_sub_;
    rclcpp::Client<studica_control::srv::SetData>::SharedPtr client_;
    rclcpp::TimerBase::SharedPtr poll_timer_;
    int print_count_{0};
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ParsecExample>());
    rclcpp::shutdown();
    return 0;
}

File: src/studica_control/src/components/examples/python/parsec_example.py
#!/usr/bin/env python3
"""Parsec — subscribe to multi-zone ToF readings (100 frames, then exit).

Run:  ros2 run studica_control parsec_example.py
Requires: studica_launch.py running, parsec enabled in params.yaml
          sensors: ["parsec"]  (name must match SENSOR below)

Topics:
  /parsec/zones     (studica_control/ParsecZoneMsg)
  /parsec/min_range (sensor_msgs/Range)
Service: /parsec/parsec_cmd (SetData)
  Commands: get_config, get_zone_distance, get_min_distance
"""
import math

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Range
from studica_control.msg import ParsecZoneMsg
from studica_control.srv import SetData

SENSOR = 'parsec'
MAX_PRINTS = 100


def _zone_cell(d: int) -> str:
    if d == -1:
        return '    off'
    if d == -2 or d < 1 or d > 4000:
        return '    ---'
    return f'{d:4d} mm'


class ParsecExample(Node):
    def __init__(self):
        super().__init__('parsec_example')
        self.create_subscription(ParsecZoneMsg, f'/{SENSOR}/zones', self.on_zones, 10)
        self.create_subscription(Range, f'/{SENSOR}/min_range', self.on_min_range, 10)
        self.client = self.create_client(SetData, f'/{SENSOR}/parsec_cmd')
        self.create_timer(5.0, self.poll_service)
        self.print_count = 0
        self.get_logger().info(f'Listening on /{SENSOR}/zones and /{SENSOR}/min_range')

    def on_zones(self, msg: ParsecZoneMsg):
        side = 8 if msg.zones == 64 else 4
        center = msg.zones // 2
        center_mm = msg.fdist[center] if center < len(msg.fdist) else -2
        self.get_logger().info(
            f'{self.print_count:3d} | zones seq={msg.seq} count={msg.zones} center={center_mm} mm'
        )
        for row in range(side):
            row_vals = msg.fdist[row * side:(row + 1) * side]
            if not row_vals:
                break
            line = '  ' + ' '.join(_zone_cell(d) for d in row_vals)
            self.get_logger().info(line)

        self.print_count += 1
        if self.print_count >= MAX_PRINTS:
            self.get_logger().info(f'printed {MAX_PRINTS} frames, exiting')
            rclpy.shutdown()

    def on_min_range(self, msg: Range):
        if math.isinf(msg.range):
            self.get_logger().warn('no valid zones in frame')
        else:
            self.get_logger().info(f'nearest valid zone: {msg.range:.3f} m')

    def poll_service(self):
        if not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn('parsec_cmd service not available')
            return
        req = SetData.Request()
        req.params = 'get_min_distance'
        future = self.client.call_async(req)
        future.add_done_callback(self._on_service_response)

    def _on_service_response(self, future):
        resp = future.result()
        if resp.success:
            self.get_logger().info(f'service get_min_distance: {resp.message} mm')
        else:
            self.get_logger().warn(f'service get_min_distance failed: {resp.message}')


def main():
    rclpy.init()
    rclpy.spin(ParsecExample())
    rclpy.shutdown()


if __name__ == '__main__':
    main()
